File: src/main/java/tn/esprit/projet/models/suivi.java
package tn.esprit.projet.models;

public class suivi {

    private int idsuivi;
    private int idutilisateur;
    private int idpsychologue;
    private String titre;
    private String description;
    private String type_suivi;

    // ===================== Constructeurs =====================
    public suivi() {}

    public suivi(int idutilisateur, int idpsychologue, String titre, String description, String type_suivi) {
        this.idutilisateur = idutilisateur;
        this.idpsychologue = idpsychologue;
        this.titre = titre;
        this.description = description;
        this.type_suivi = type_suivi;
    }

    // ===================== Getters MAJUSCULE (obligatoires pour PropertyValueFactory) =====================
    public int getIdsuivi()        { return idsuivi; }
    public int getIdutilisateur()  { return idutilisateur; }
    public int getIdpsychologue()  { return idpsychologue; }
    public String getTitre()       { return titre; }
    public String getDescription() { return description; }
    public String getType_suivi()  { return type_suivi; }

    // ===================== Getters minuscule (pour service et controller) =====================
    public int getidsuivi()        { return idsuivi; }
    public int getidutilisateur()  { return idutilisateur; }
    public int getidpsychologue()  { return idpsychologue; }
    public String gettitre()       { return titre; }
    public String getdescription() { return description; }
    public String gettype_suivi()  { return type_suivi; }

    // ===================== Setters =====================
    public void setidsuivi(int idsuivi)             { this.idsuivi = idsuivi; }
    public void setidutilisateur(int idutilisateur) { this.idutilisateur = idutilisateur; }
    public void setidpsychologue(int idpsychologue) { this.idpsychologue = idpsychologue; }
    public void settitre(String titre)              { this.titre = titre; }
    public void setdescription(String description)  { this.description = description; }
    public void settype_suivi(String type_suivi)    { this.type_suivi = type_suivi; }

    @Override
    public String toString() {
        return "suivi{id=" + idsuivi + ", titre='" + titre + "', type_suivi='" + type_suivi + "'}";
    }
}

File: src/Entity/Suivi.php
<?php
namespace App\Entity;

use App\Repository\SuiviRepository;
use Doctrine\ORM\Mapping as ORM;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;

#[ORM\Entity(repositoryClass: SuiviRepository::class)]
#[ORM\Table(name: 'suivis')]
class Suivi
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'id_suivi')]
    private ?int $idsuivi = null;

    #[ORM\ManyToOne(targetEntity: User::class)]
    #[ORM\JoinColumn(name: 'id_utilisateur', referencedColumnName: 'id', nullable: false, onDelete: 'CASCADE')]
    private ?User $utilisateur = null;

    #[ORM\ManyToOne(targetEntity: User::class)]
    #[ORM\JoinColumn(name: 'id_psychologue', referencedColumnName: 'id', nullable: true, onDelete: 'SET NULL')]
    private ?User $psychologue = null;

    #[ORM\Column(name: 'titre', length: 255)]
    private ?string $titre = null;

    #[ORM\Column(name: 'description', type: 'text', nullable: true)]
    private ?string $description = null;

    #[ORM\Column(name: 'date_creation', type: 'datetime', nullable: true)]
    private ?\DateTimeInterface $date_creation = null;

    #[ORM\Column(name: 'date_modification', type: 'datetime', nullable: true)]
    private ?\DateTimeInterface $date_modification = null;

    #[ORM\Column(name: 'type_suivi', length: 50)]
    private ?string $type_suivi = null;

    #[ORM\OneToMany(mappedBy: 'suivi', targetEntity: Objectif::class, cascade: ['persist', 'remove'])]
    private Collection $objectifs;

    public function __construct()
    {
        $this->objectifs = new ArrayCollection();
        $this->date_creation = new \DateTime();
        $this->date_modification = new \DateTime();
    }

    public function getIdsuivi(): ?int { return $this->idsuivi; }

    public function getUtilisateur(): ?User { return $this->utilisateur; }
    public function setUtilisateur(?User $utilisateur): self { $this->utilisateur = $utilisateur; return $this; }

    // Compat backward : retourne l'ID brut si besoin
    public function getIdutilisateur(): ?int { return $this->utilisateur?->getId(); }
    public function setIdutilisateur(int $id): self { return $this; } // no-op, utiliser setUtilisateur()

    public function getPsychologue(): ?User { return $this->psychologue; }
    public function setPsychologue(?User $psychologue): self { $this->psychologue = $psychologue; return $this; }

    public function getIdpsychologue(): ?int { return $this->psychologue?->getId(); }
    public function setIdpsychologue(int $id): self { return $this; } // no-op

    public function getTitre(): ?string { return $this->titre; }
    public function setTitre(string $titre): self { $this->titre = $titre; return $this; }

    public function getDescription(): ?string { return $this->description; }
    public function setDescription(?string $description): self { $this->description = $description; return $this; }

    public function getDateCreation(): ?\DateTimeInterface { return $this->date_creation; }
    public function setDateCreation(\DateTimeInterface $date_creation): self { $this->date_creation = $date_creation; return $this; }

    public function getDateModification(): ?\DateTimeInterface { return $this->date_modification; }
    public function setDateModification(\DateTimeInterface $date_modification): self { $this->date_modification = $date_modification; return $this; }

    public function getTypeSuivi(): ?string { return $this->type_suivi; }
    public function setTypeSuivi(string $type_suivi): self { $this->type_suivi = $type_suivi; return $this; }

    public function getObjectifs(): Collection { return $this->objectifs; }

    public function addObjectif(Objectif $objectif): self
    {
        if (!$this->objectifs->contains($objectif)) {
            $this->objectifs->add($objectif);
            $objectif->setSuivi($this);
        }
        return $this;
    }

    public function removeObjectif(Objectif $objectif): self
    {
        if ($this->objectifs->removeElement($objectif)) {
            if ($objectif->getSuivi() === $this) {
                $objectif->setSuivi(null);
            }
        }
        return $this;
    }

    public function __toString(): string
    {
        return $this->titre ?? 'Suivi #' . $this->idsuivi;
    }
}
